File: CS/Q260885/MyGridControl.cs
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid.Drawing;
using DevExpress.Utils.Drawing;
using DevExpress.Utils;
using System.Drawing;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace DXSample {
    public class MyGridControl :GridControl {
        public MyGridControl() : base() { }

        protected override void RegisterAvailableViewsCore(InfoCollection collection) {
            base.RegisterAvailableViewsCore(collection);
            collection.Add(new MyGridViewInfoRegistrator());
        }
    }

    public class MyGridView :GridView {
        public MyGridView() : base() { }
        public MyGridView(GridControl grid) : base(grid) { }

        internal const string MyGridViewName = "MyGridView";
        protected override string ViewName { get { return MyGridViewName; } }
    }

    public class MyGridViewInfoRegistrator :GridInfoRegistrator {
        public MyGridViewInfoRegistrator() : base() { }

        public override string ViewName { get { return MyGridView.MyGridViewName; } }

        public override BaseView CreateView(GridControl grid) {
            return new MyGridView(grid);
        }

        public override BaseViewPainter CreatePainter(BaseView view) {
            return new MyGridViewPainter((MyGridView)view);
        }
    }

    public class MyGridViewPainter :GridPainter {
        public MyGridViewPainter(MyGridView view) : base(view) { }

        protected override void DrawRegularRow(GridViewDrawArgs e, GridDataRowInfo ri) {
            base.DrawRegularRow(e, ri);
            if (ri.IsMasterRow && ri.MasterRowExpanded) {
                GridView detailView = View.GetDetailView(ri.RowHandle, 0) as GridView;
                if (detailView == null) return;

                GridCellInfo cell = ri.Cells[View.VisibleColumns[0]];
                int center;
                if (cell != null) center = cell.CellButtonRect.Left + cell.CellButtonRect.Width / 2;
                else center = ri.DetailIndentBounds.Left + ri.DetailIndentBounds.Width / 2 ;

                GridViewInfo detailViewInfo = (GridViewInfo)detailView.GetViewInfo();
                int level = 0;
                Point p1;
                Point p2;
                foreach (GridRowInfo rowInfo in detailViewInfo.RowsInfo) {
                    if (detailView.IsRowVisible(rowInfo.RowHandle) != RowVisibleState.Visible) continue;
                    level = rowInfo.Bounds.Top + rowInfo.Bounds.Height / 2;
                    p1 = new Point(center, level);
                    p2 = new Point(ri.DetailIndentBounds.Right, level);
                    e.Graphics.DrawLine(Pens.Black, p1, p2);
                }

                p1 = new Point(center, ri.DetailIndentBounds.Top);
                p2 = new Point(center, level);
                e.Graphics.DrawLine(Pens.Black, p1, p2);
            }
        }
    }
}

File: VB/Q260885/MyGridControl.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid.Drawing
Imports DevExpress.Utils.Drawing
Imports DevExpress.Utils
Imports System.Drawing
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace DXSample
	Public Class MyGridControl
		Inherits GridControl
		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Sub RegisterAvailableViewsCore(ByVal collection As InfoCollection)
			MyBase.RegisterAvailableViewsCore(collection)
			collection.Add(New MyGridViewInfoRegistrator())
		End Sub
	End Class

	Public Class MyGridView
		Inherits GridView
		Public Sub New()
			MyBase.New()
		End Sub
		Public Sub New(ByVal grid As GridControl)
			MyBase.New(grid)
		End Sub

		Friend Const MyGridViewName As String = "MyGridView"
		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return MyGridViewName
			End Get
		End Property
	End Class

	Public Class MyGridViewInfoRegistrator
		Inherits GridInfoRegistrator
		Public Sub New()
			MyBase.New()
		End Sub

		Public Overrides ReadOnly Property ViewName() As String
			Get
				Return MyGridView.MyGridViewName
			End Get
		End Property

		Public Overrides Function CreateView(ByVal grid As GridControl) As BaseView
			Return New MyGridView(grid)
		End Function

		Public Overrides Function CreatePainter(ByVal view As BaseView) As BaseViewPainter
			Return New MyGridViewPainter(CType(view, MyGridView))
		End Function
	End Class

	Public Class MyGridViewPainter
		Inherits GridPainter
		Public Sub New(ByVal view As MyGridView)
			MyBase.New(view)
		End Sub

		Protected Overrides Sub DrawRegularRow(ByVal e As GridViewDrawArgs, ByVal ri As GridDataRowInfo)
			MyBase.DrawRegularRow(e, ri)
			If ri.IsMasterRow AndAlso ri.MasterRowExpanded Then
				Dim detailView As GridView = TryCast(View.GetDetailView(ri.RowHandle, 0), GridView)
				If detailView Is Nothing Then
					Return
				End If

				Dim cell As GridCellInfo = ri.Cells(View.VisibleColumns(0))
				Dim center As Integer
				If cell IsNot Nothing Then
					center = cell.CellButtonRect.Left + cell.CellButtonRect.Width \ 2
				Else
					center = ri.DetailIndentBounds.Left + ri.DetailIndentBounds.Width \ 2
				End If

				Dim detailViewInfo As GridViewInfo = CType(detailView.GetViewInfo(), GridViewInfo)
				Dim level As Integer = 0
				Dim p1 As Point
				Dim p2 As Point
				For Each rowInfo As GridRowInfo In detailViewInfo.RowsInfo
					If detailView.IsRowVisible(rowInfo.RowHandle) <> RowVisibleState.Visible Then
						Continue For
					End If
					level = rowInfo.Bounds.Top + rowInfo.Bounds.Height \ 2
					p1 = New Point(center, level)
					p2 = New Point(ri.DetailIndentBounds.Right, level)
					e.Graphics.DrawLine(Pens.Black, p1, p2)
				Next rowInfo

				p1 = New Point(center, ri.DetailIndentBounds.Top)
				p2 = New Point(center, level)
				e.Graphics.DrawLine(Pens.Black, p1, p2)
			End If
		End Sub
	End Class
End Namespace